mypy/server/update.py

```diff
@@ -174,70 +211,85 @@ def get_all_dependencies(manager: BuildManager, graph: Dict[str, State],
-
-    Return the new ASTs for the changed modules and the entire build graph.
+def update_single_isolated(module: str,
+                           path: str,
+                           manager: BuildManager,
+                           graph: Dict[str, State]) -> Tuple[str,
+                                                             Optional[MypyFile],
+                                                             Graph,
+                                                             List[Tuple[str, str]]]:
+    """Build a new version of one changed module only.
+
+    Don't propagate changes to elsewhere in the program. Raise CompleError on
+    encountering a blocking error.
+
+    Args:
+        module: Changed module (modified, created or deleted)
+        path: Path of the changed module
+        manager: Build manager
+        graph: Build graph
+
+    Returns:
+        A 4-tuple with these items:
+
+        - Id of the changed module (can be different from the argument)
+        - New AST for the changed module (None if module was deleted)
+        - The entire build graph
+        - Remaining changed modules that are not processed yet as (module id, path)
+          tuples (non-empty if the original changed module imported other new
+          modules)
     """
-    # TODO: Handle multiple changed modules per step
-    assert len(changed_modules) == 1
-    id, path = changed_modules[0]
-    if id in manager.modules:
-        path1 = os.path.normpath(path)
-        path2 = os.path.normpath(manager.modules[id].path)
-        assert path1 == path2, '%s != %s' % (path1, path2)
+    if module in manager.modules:
+        assert_equivalent_paths(path, manager.modules[module].path)
 
     old_modules = dict(manager.modules)
-
-    sources = get_sources(graph, changed_modules)
-    changed_set = {id for id, _ in changed_modules}
-
-    invalidate_stale_cache_entries(manager.saved_cache, changed_modules)
+    sources = get_sources(graph, [(module, path)])
+    invalidate_stale_cache_entries(manager.saved_cache, [(module, path)])
 
     if not os.path.isfile(path):
-        graph = delete_module(id, graph, manager)
-        return {id: None}, graph
+        graph = delete_module(module, graph, manager)
+        return module, None, graph, []
 
     old_graph = graph
     manager.missing_modules = set()
     graph = load_graph(sources, manager)
 
     # Find any other modules brought in by imports.
-    for st in graph.values():
-        if st.id not in old_graph and st.id not in changed_set:
-            changed_set.add(st.id)
-            assert st.path
-            changed_modules.append((st.id, st.path))
-    # TODO: Handle multiple changed modules per step
-    assert len(changed_modules) == 1, changed_modules
+    changed_modules = get_all_changed_modules(module, path, old_graph, graph)
+    # If there are multiple modules to process, only process one of them and return the
+    # remaining ones to the caller.
```

Comment:
Maybe note that you process the last one in the list first? This is significant, because the first one in the list is guaranteed (by the implementation of `get_all_changed_modules()`) to be the original module.